tools/torchci/test_insights/file_report_generator.py

```diff
@@ -0,0 +1,690 @@
+    def load_test_owners(self) -> List[Dict[str, Any]]:
+        """Load the test owner labels JSON file from S3"""
+        S3_URL = "https://ossci-metrics.s3.us-east-1.amazonaws.com/test_owner_labels/test_owner_labels.json.gz"
+        print(f"Fetching test owner labels from S3: {S3_URL}")
+        with urllib.request.urlopen(S3_URL) as response:
+            compressed_data = response.read()
+        decompressed_data = gzip.decompress(compressed_data)
+        test_owners = []
+        for line in decompressed_data.decode("utf-8").splitlines():
+            if line.strip():
+                test_owners.append(json.loads(line))
+        return test_owners
+
+    def get_runner_cost(self, runner_label: str) -> float:
+        """Get the cost per hour for a given runner"""
+        runner_costs = self.load_runner_costs()
+        if runner_label.startswith("lf."):
+            runner_label = runner_label[3:]
+        return runner_costs.get(runner_label, 0.0)
+
+    def _get_first_suitable_sha(self, shas: list[dict[str, Any]]) -> Optional[str]:
+        """Get the first suitable SHA from a list of SHAs."""
+        lens = []
+        for sha in shas:
+            head_sha = sha["head_sha"]
+            test_data = self._get_invoking_file_test_data_for_sha(head_sha)
+
+            has_no_job_name = False
+            for entry in test_data:
+                if "NoJobName" in entry.get("short_job_name", ""):
+                    has_no_job_name = True
+                    break
+            if has_no_job_name:
+                print(f"Has entries with no job name for {head_sha}")
+                continue
+
+            lens.append((head_sha, len(test_data)))
+            del test_data
+
+            if len(lens) > 1:
+                lens.sort(key=lambda x: x[1], reverse=True)
+                if abs(lens[0][1] - lens[1][1]) * 2 / (lens[0][1] + lens[1][1]) < 0.1:
+                    print(f"Using SHA {lens[0][0]} with {lens[0][1]} entries")
+                    return lens[0][0]
+        return None
+
+    def find_suitable_sha(self, date: str) -> Optional[str]:
+        """
+        Auto-select suitable SHA from PyTorch main branch for a given date.
+        Usage:
+        - Provide a date to select a SHA from that day
+        Criteria:
+        - SHA is from main branch
+        - Workflow jobs are successful (green)
+        - S3 test data is available
+        - All test entries have job names
+        """
+
+        print("Searching for suitable SHAs from PyTorch main branch...")
+
+        params = {
+            "start_date": date + " 00:00:00",
+            "stop_date": date + " 23:59:59",
+        }
+
+        # Single query with conditional logic using CASE expressions
+        query = """
+        SELECT
+            w.head_sha,
+            toUnixTimestamp(w.head_commit.'timestamp') as push_date
+        FROM default.workflow_run w
+        WHERE w.head_branch = 'main'
+            AND w.repository.full_name = 'pytorch/pytorch'
```

Comment:
Do you think we want to make this parameter configurable on the workflow?  It's easier to find it there